### PR TITLE
feat: add tuple support and exam_cb for callbacks

### LINKED ISSUE #635: Repeated iterations in callback
I tried using the `callback` to provide progress indicated. But using a simple `print` functions, I am quite confused by the output: Why do I get repeated output for the same step and back, as soon as the progress halts?

Example code:
```python
from interpret.glassbox import ExplainableBoostingRegressor
from interpret.utils import make_synthetic
from sklearn.model_selection import train_test_split

seed = 42

X, y, names, types = make_synthetic(classes=None, n_samples=50000, missing=False, seed=seed)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20, random_state=seed)


def logging_callback(bag_idx: int, n_steps: int, has_progressed: bool, best_score: float) -> bool:
    if n_steps % 10000 == 0:
        print(f"{bag_idx=:2}, {n_steps=:5}, {has_progressed=!s:>5}, {best_score=:.6}")
    return False


ebm = ExplainableBoostingRegressor(names, types, interactions=3, smoothing_rounds=5000, reg_alpha=10.0, callback=logging_callback)
ebm.fit(X_train, y_train)
```

Some output:
```
bag_idx= 6, n_steps=100000, has_progressed= True, best_score=0.313481
bag_idx= 6, n_steps=100000, has_progressed=False, best_score=0.313481
bag_idx= 6, n_steps=100000, has_progressed=False, best_score=0.313481
bag_idx= 6, n_steps=100000, has_progressed=False, best_score=0.313481
bag_idx= 6, n_steps=100000, has_progressed=False, best_score=0.313481
bag_idx= 6, n_steps=100000, has_progressed=False, best_score=0.313481
bag_idx= 6, n_steps=100000, has_progressed=False, best_score=0.313481
bag_idx= 6, n_steps=100000, has_progressed=False, best_score=0.313481
bag_idx= 6, n_steps=100000, has_progressed=False, best_score=0.313481
bag_idx= 6, n_steps=100000, has_progressed=False, best_score=0.313481
bag_idx= 6, n_steps=100000, has_progressed=False, best_score=0.313481
```

### PR DESCRIPTION
**Pull request by ugbotueferhire**
Fixes #635 Callback tuple support and exam callback

## Description

This PR implements Phase 2 of the callback redesign for EBM training.

Phase 1 changed the callback API to keyword-only arguments and ensured the
progress callback only fires on progressing boosting steps. That work was
merged in #662.

This PR builds on top of that by allowing `callback` to accept either:

- a single progress callback
- a single examination callback
- a tuple containing one progress callback and one examination callback

The supported callback signatures are:

```python
def progress_cb(*, bag, stage, step, term, metric):
    ...

def exam_cb(*, bag, stage, step, term, gain):
    ...
```

## Why

The original callback parameter was carrying more than one kind of signal.
As discussed in the review for #662, there are really two different callback
use cases:

- a progress callback that reports accepted boosting progress
- an examination callback that reports when a term has been examined and what
  gain was computed for it

Splitting those concepts makes the API clearer while still keeping a single
public `callback` parameter.

## What Changed

- Added callback normalization and validation in `_ebm.py`
- Classified callbacks by signature using `metric` vs `gain`
- Allowed `callback` to be either a callable or a tuple of callables
- Enforced at most one progress callback and one examination callback
- Added an `exam_callback` hook in `_boost.py` at the point where `avg_gain`
  is computed
- Kept the existing progress callback behavior after accepted term updates
- Allowed either callback type to stop training early by returning `True`
- Updated callback parameter docstrings for EBM model, classifier, and regressor

## Validation Rules

This PR raises a `ValueError` for invalid callback configurations such as:

- empty callback tuples
- more than one progress callback
- more than one examination callback
- callbacks that do not match either supported signature

## Tests

Added or covered tests for:

- tuple callback support calling both callback types
- valid finite `gain` values in the examination callback
- early termination from the examination callback
- invalid tuple configurations
- invalid callback signatures

Verified with:

```bash
python -m pytest tests/glassbox/ebm/test_callback.py -q
python -m pytest tests/glassbox/ebm/test_ebm.py::test_callbacks_short tests/glassbox/ebm/test_ebm.py::test_callbacks_long tests/glassbox/ebm/test_merge_ebms.py::test_merge_ebms_callback_is_none -q
```

## Notes

- This preserves the Phase 1 progress callback behavior introduced in #662.
- The final intercept-only pass still does not invoke training callbacks.

### CODE REVIEW

**Comment by ugbotueferhire** [comment]
@paulbkoch conflicts has been resolved

File: python/interpret-core/interpret/glassbox/_ebm.py [hunk 1]
```diff
@@ -78,6 +79,88 @@
...
+
+    callbacks = callback if isinstance(callback, tuple) else (callback,)
+    if len(callbacks) == 0:
+        msg = "callback tuple cannot be empty"
+        _log.error(msg)
+        raise ValueError(msg)
+    if len(callbacks) > 2:
+        msg = "callback tuple can contain at most one progress callback and one examination callback"
+        _log.error(msg)
+        raise ValueError(msg)
```
Comment (paulbkoch) [inline python/interpret-core/interpret/glassbox/_ebm.py]: We can get rid of these checks.  Empty callback tuple should be allowed and return None, None which it does in the code below already.  And if callbacks has more than two values then it will be detected by the duplicate checks below.

File: python/interpret-core/interpret/glassbox/_ebm.py [hunk 2]
```diff
@@ -78,6 +79,88 @@
...
+
+    try:
+        signature.bind(**{name: None for name in required_names})
+    except TypeError as exc:
+        msg = f"callback must be callable with keyword arguments {required_names}"
+        _log.error(msg)
+        raise ValueError(msg) from exc
+
+    return "progress" if has_metric else "exam"
```
Comment (paulbkoch) [inline python/interpret-core/interpret/glassbox/_ebm.py]: This is probably simpler and more expandable as something like:

callback_types = {"progress" : {"bag", "stage", "step", "term", "metric"}, "exam": {"bag", "stage", "step", "term", "gain"}}

param_names = set(inspect.signature(callback).parameters)

for name, params in callback_types.items():
    if params == param_names:
         return name

raise something

File: python/interpret-core/interpret/glassbox/_ebm_core/_boost.py [hunk 3]
```diff
@@ -264,6 +265,19 @@ def boost(
                                 # penalize nominals a bit because they benefit from sorting categories
                                 avg_gain *= gain_scale
```
Comment (paulbkoch) [inline python/interpret-core/interpret/glassbox/_ebm_core/_boost.py:267]: should check 

                            if stop_flag is not None and stop_flag[0]:
                                break

here

File: python/interpret-core/interpret/glassbox/_ebm_core/_boost.py
Comment (paulbkoch) [inline python/interpret-core/interpret/glassbox/_ebm_core/_boost.py:399]: This one can be removed since we're checking before both callbacks

File: python/interpret-core/interpret/glassbox/_ebm.py [hunk 4]
```diff
@@ -3431,13 +3519,13 @@ def __init__(
         # Boosting
         learning_rate: float = 0.02,
         greedy_ratio: float | None = 10.0,
-        cyclic_progress: bool | float = False,
+        cyclic_progress: bool | float | int = False,  # noqa: PYI041
```
Comment (paulbkoch) [inline python/interpret-core/interpret/glassbox/_ebm.py]: The float type hint includes int, so remove int and the noqa

File: python/interpret-core/interpret/glassbox/_ebm.py [hunk 5]
```diff
@@ -3755,13 +3843,13 @@ def __init__(
         # Boosting
         learning_rate: float = 0.015,
         greedy_ratio: float | None = 10.0,
-        cyclic_progress: bool | float = False,
+        cyclic_progress: bool | float | int = False,  # noqa: PYI041
```
Comment (paulbkoch) [inline python/interpret-core/interpret/glassbox/_ebm.py]: remove int since float includes the int type hint, and also remove noqa

File: python/interpret-core/interpret/glassbox/_ebm.py [hunk 6]
```diff
@@ -4085,13 +4173,13 @@ def __init__(
         # Boosting
         learning_rate: float = 0.04,
         greedy_ratio: float | None = 10.0,
-        cyclic_progress: bool | float = False,
+        cyclic_progress: bool | float | int = False,  # noqa: PYI041
```
Comment (paulbkoch) [inline python/interpret-core/interpret/glassbox/_ebm.py]: same as above.  remove int and noqa

File: python/interpret-core/tests/glassbox/ebm/test_callback.py [hunk 7]
```diff
@@ -223,3 +248,273 @@ def __call__(self, *, bag, stage, step, term, metric):
...
+
+@pytest.mark.parametrize(
+    "callback, message",
+    [
+        ((RecordingCallback(), RecordingCallback()), "more than one progress callback"),
+        (
+            (ExamRecordingCallback(), ExamRecordingCallback()),
+            "more than one examination callback",
+        ),
+        (tuple(), "cannot be empty"),
```
Comment (paulbkoch) [inline python/interpret-core/tests/glassbox/ebm/test_callback.py]: empty tuple should be valid as a no callbacks case

File: python/interpret-core/tests/glassbox/ebm/test_ebm.py [hunk 8]
```diff
@@ -249,6 +249,32 @@ def test_ebm_sweep():
...
+        interactions=interactions,
+        outer_bags=1,
+        max_rounds=10,
+        n_jobs=1,
+    )
+
+    with pytest.raises(ValueError, match=message):
+        ebm.fit(X, y)
+
+
```
Comment (paulbkoch) [inline python/interpret-core/tests/glassbox/ebm/test_ebm.py]: this test seems out of scope for the callback change

**Comment by paulbkoch** [comment]
Looks great!  Thank you @ugbotueferhire.
